In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score



In [2]:
df = pd.read_csv("emails.csv")
df.head(20)

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0    Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1    Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2    Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3    Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4    Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   
5    Email 6    4   5    1    4    2   3   45    1    0  ...         0    0   
6    Email 7    5   3    1    3    2   1   37    0    0  ...         0    0   
7    Email 8    0   2    2    3    1   2   21    6    0  ...         0    0   
8    Email 9    2   2    3    0    0   1   18    0    0  ...         0    0   
9   Email 10    4   4   35    0    1   0   49    1   16  ...         0    0   
10  Email 11   22  14    2    9    2   2  104    0    2  ...         0    0   
11  Email 12   33  28   27   11   10  12  173    6   12  ...         0    0   
12  Email 13   27  17    3    7    5   8  106    3    0  ...         0    0   
13  Email 14    4   5    7    1    5   1   37    1    3  ...         0    0   
14  Email 15    2   4    6    0    3   1   16    0    3  ...         0    0   
15  Email 16    6   2    1    0    2   0   36    3    1  ...         0    0   
16  Email 17    3   1    2    2    0   1   17    0    0  ...         0    0   
17  Email 18   36  21    6   14    7  17  194   25    5  ...         0    0   
18  Email 19    1   3    1    0    2   0   14    0    0  ...         0    0   
19  Email 20    3   4   11    0    4   2   32    1    5  ...         0    0   

    valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0        0    0               0         0         0   0    0           0  
1        0    0               0         0         0   1    0           0  
2        0    0               0         0         0   0    0           0  
3        0    0               0         0         0   0    0           0  
4        0    0               0         0         0   1    0           0  
5        0    0               0         0         0   0    0           1  
6        0    0               0         0         0   0    0           0  
7        0    0               0         0         0   1    0           1  
8        0    0               0         0         0   0    0           0  
9        0    0               0         0         0   0    0           0  
10       0    0               0         0         0   1    0           0  
11       0    0               0         0         0   5    0           0  
12       0    0               0         0         0   4    0           0  
13       0    0               0         0         0   2    0           0  
14       0    0               0         0         0   1    0           0  
15       0    0               0         0         0   0    0           0  
16       0    0               0         0         0   1    0           1  
17       0    0               0         0         0   3    0           1  
18       0    0               0         0         0   0    0           0  
19       0    0               0         0         0   1    0           0  

[20 rows x 3002 columns]

In [3]:
df.isnull().sum()

Email No.     0
the           0
to            0
ect           0
and           0
             ..
military      0
allowing      0
ff            0
dry           0
Prediction    0
Length: 3002, dtype: int64

In [4]:
df.describe()

the           to          ect          and          for  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000   
mean      6.640565     6.188128     5.143852     3.075599     3.124710   
std      11.745009     9.534576    14.101142     6.045970     4.680522   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       0.000000     1.000000     1.000000     0.000000     1.000000   
50%       3.000000     3.000000     1.000000     1.000000     2.000000   
75%       8.000000     7.000000     4.000000     3.000000     4.000000   
max     210.000000   132.000000   344.000000    89.000000    47.000000   

                of            a          you          hou           in  ...  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  ...   
mean      2.627030    55.517401     2.466551     2.024362    10.600155  ...   
std       6.229845    87.574172     4.314444     6.967878    19.281892  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000    12.000000     0.000000     0.000000     1.000000  ...   
50%       1.000000    28.000000     1.000000     0.000000     5.000000  ...   
75%       2.000000    62.250000     3.000000     1.000000    12.000000  ...   
max      77.000000  1898.000000    70.000000   167.000000   223.000000  ...   

          connevey          jay       valued          lay  infrastructure  \
count  5172.000000  5172.000000  5172.000000  5172.000000     5172.000000   
mean      0.005027     0.012568     0.010634     0.098028        0.004254   
std       0.105788     0.199682     0.116693     0.569532        0.096252   
min       0.000000     0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000     0.000000        0.000000   
50%       0.000000     0.000000     0.000000     0.000000        0.000000   
75%       0.000000     0.000000     0.000000     0.000000        0.000000   
max       4.000000     7.000000     2.000000    12.000000        3.000000   

          military     allowing           ff          dry   Prediction  
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  
mean      0.006574     0.004060     0.914733     0.006961     0.290023  
std       0.138908     0.072145     2.780203     0.098086     0.453817  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     1.000000     0.000000     1.000000  
max       4.000000     3.000000   114.000000     4.000000     1.000000  

[8 rows x 3001 columns]

In [5]:
df.corr()

the        to       ect       and       for        of  \
the         1.000000  0.852715  0.337249  0.841200  0.784112  0.796397   
to          0.852715  1.000000  0.375480  0.825474  0.781971  0.752722   
ect         0.337249  0.375480  1.000000  0.272863  0.369777  0.178028   
and         0.841200  0.825474  0.272863  1.000000  0.751287  0.809665   
for         0.784112  0.781971  0.369777  0.751287  1.000000  0.681457   
...              ...       ...       ...       ...       ...       ...   
military    0.129466  0.091639 -0.007690  0.084147  0.067151  0.073004   
allowing    0.127019  0.120059  0.004368  0.124766  0.121057  0.108786   
ff          0.341878  0.406666  0.141460  0.400225  0.301074  0.444252   
dry         0.051021  0.071388  0.002492  0.042484  0.038126  0.026403   
Prediction -0.004421  0.055277 -0.120782  0.114364 -0.003101  0.197234   

                   a       you       hou        in  ...  connevey       jay  \
the         0.784451  0.471392  0.303621  0.845670  ...  0.008926  0.075479   
to          0.896466  0.508513  0.347993  0.881759  ...  0.013250  0.101247   
ect         0.400009  0.155783  0.974152  0.298387  ...  0.134339  0.031431   
and         0.815196  0.476764  0.235953  0.874276  ...  0.005151  0.104454   
for         0.744098  0.495852  0.329051  0.762659  ...  0.022168  0.041775   
...              ...       ...       ...       ...  ...       ...       ...   
military    0.111685  0.006498  0.005429  0.120620  ... -0.002249 -0.002979   
allowing    0.105358  0.082757 -0.000966  0.138099  ... -0.002675 -0.003543   
ff          0.464473  0.195058  0.114210  0.448303  ...  0.005403  0.073690   
dry         0.093822  0.028883  0.000601  0.077751  ... -0.003373  0.035028   
Prediction  0.107776  0.130293 -0.128340  0.154055  ... -0.030375 -0.031694   

              valued       lay  infrastructure  military  allowing        ff  \
the         0.225586  0.223426        0.101768  0.129466  0.127019  0.341878   
to          0.232847  0.255793        0.093322  0.091639  0.120059  0.406666   
ect         0.046080  0.061550        0.004393 -0.007690  0.004368  0.141460   
and         0.272963  0.253440        0.151980  0.084147  0.124766  0.400225   
for         0.236213  0.213631        0.134469  0.067151  0.121057  0.301074   
...              ...       ...             ...       ...       ...       ...   
military    0.043408  0.104297        0.041300  1.000000  0.055227  0.049524   
allowing   -0.005130  0.018550        0.276001  0.055227  1.000000  0.096212   
ff          0.130356  0.164296        0.114092  0.049524  0.096212  1.000000   
dry        -0.006468  0.018939       -0.003137  0.010835 -0.003995  0.049690   
Prediction  0.098775  0.064315        0.038161  0.064850  0.011279  0.135479   

                 dry  Prediction  
the         0.051021   -0.004421  
to          0.071388    0.055277  
ect         0.002492   -0.120782  
and         0.042484    0.114364  
for         0.038126   -0.003101  
...              ...         ...  
military    0.010835    0.064850  
allowing   -0.003995    0.011279  
ff          0.049690    0.135479  
dry         1.000000   -0.006260  
Prediction -0.006260    1.000000  

[3001 rows x 3001 columns]

# Creating the NB Model

In this project we are clasifying mails typed in by the user as either 'Spam' or 'Not Spam'. Our original dataset was a folder of 5172 text files containing the emails.

Now let us understand why we have separated the words from the mails. This is because, this is a text-classification problem. When a spam classifier looks at a mail, it searches for potential words that it has seen in the previous spam emails. If it finds a majority of those words, then it labels it as 'Spam'. **Why did I say majority ? -->**

*CASE 1* : suppose let's take a word 'Greetings'. Say, it is present in both 'Spam' and 'Not Spam' mails.

*CASE 2* : Let's consider a word 'lottery'.Say, it is present in only 'Spam' mails.

*CASE 3* : Let's consider a word 'cheap'. Say, it is present only in spam.

If now we get a test email, and it contains all the three words metioned above, there's high probability that it is a 'Spam' mail.

The most effective algorithm for text-classification problems is the Naive Bayes algorithm, that works on the classic Bayes' theorem. This theorem works on every individual word in the test data to make predictions(the conditional probability with higher probability is the predicted result). 

________________________________________________________________________________________________________________________

Say, our test email(S)is,*"You have won a lottery"*

**HOW NAIVE BAYES WORKS ON THIS DATA -->**

P(S) = P('You') * P('have') * P('won') * P('a') * P('lottery') ____ 1

Therefore, P(S|Spam) = P('You'|Spam) * P('have'|Spam) * P('won'|Spam) * P('a'|Spam) * P('lottery'|Spam) ____ 2 

Same calculation for P(S|Not_Spam) ____ 3

If 2 > 3, then 'Spam' Else, 'Not_Spam'.

**WHAT IF THE PROBABILITY IS ZERO ?** Here comes the concept of Laplace Smoothing, where P(words) = (word_count + 1)/(total_no_of_words + no_of_unique_words)

________________________________________________________________________________________________________________________

Here, we'll work on the existing Multinomial Naive Bayes classifier (under scikit-learn). To further understand how well Naive Bayes works for text-classification, we'll use another standard classifier, SVC, to see how the two models perform.

**WILL ENSEMBLE MODELS WORKS BETTER ?** Let us see. We will use Random Forests to compare.

In [6]:
X = df.iloc[:,1:3001]
X

the  to  ect  and  for  of    a  you  hou  in  ...  enhancements  \
0       0   0    1    0    0   0    2    0    0   0  ...             0   
1       8  13   24    6    6   2  102    1   27  18  ...             0   
2       0   0    1    0    0   0    8    0    0   4  ...             0   
3       0   5   22    0    5   1   51    2   10   1  ...             0   
4       7   6   17    1    5   2   57    0    9   3  ...             0   
...   ...  ..  ...  ...  ...  ..  ...  ...  ...  ..  ...           ...   
5167    2   2    2    3    0   0   32    0    0   5  ...             0   
5168   35  27   11    2    6   5  151    4    3  23  ...             0   
5169    0   0    1    1    0   0   11    0    0   1  ...             0   
5170    2   7    1    0    2   1   28    2    0   8  ...             0   
5171   22  24    5    1    6   5  148    8    2  23  ...             0   

      connevey  jay  valued  lay  infrastructure  military  allowing  ff  dry  
0            0    0       0    0               0         0         0   0    0  
1            0    0       0    0               0         0         0   1    0  
2            0    0       0    0               0         0         0   0    0  
3            0    0       0    0               0         0         0   0    0  
4            0    0       0    0               0         0         0   1    0  
...        ...  ...     ...  ...             ...       ...       ...  ..  ...  
5167         0    0       0    0               0         0         0   0    0  
5168         0    0       0    0               0         0         0   1    0  
5169         0    0       0    0               0         0         0   0    0  
5170         0    0       0    0               0         0         0   1    0  
5171         0    0       0    0               0         0         0   0    0  

[5172 rows x 3000 columns]

In [7]:
Y = df.iloc[:,-1].values
Y

array([0, 0, 0, ..., 1, 1, 0])

In [8]:
train_x,test_x,train_y,test_y = train_test_split(X,Y,test_size = 0.25)

# Naive Bayes

In [9]:
mnb = MultinomialNB(alpha=1.9)         # alpha by default is 1. alpha must always be > 0. 
# alpha is the '1' in the formula for Laplace Smoothing (P(words))
mnb.fit(train_x,train_y)
y_pred1 = mnb.predict(test_x)
print("Accuracy Score for Naive Bayes : ", accuracy_score(y_pred1,test_y))

Accuracy Score for Naive Bayes :  0.9381283836040216
